File: sloth/src/main/java/com/xiansenliu/sloth/PuppetFragment.java
package com.xiansenliu.sloth;

import android.annotation.SuppressLint;
import android.app.Fragment;
import android.content.pm.PackageManager;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Author       xinliu
 * Date         6/24/17
 * Time         3:08 PM
 * Used to request permissions and get the results
 */

public final class PuppetFragment extends Fragment {
    public static final int GO_SETTING_CODE = 9999;
    private Request mRequest;

    @SuppressLint("NewApi")
    public void request(Request request) {
        setRequest(request);
        String[] permissions = mRequest.getPermissions();
        int requestCode = mRequest.getRequestCode();

        boolean granted = Utils.arePermissionsGranted(getActivity(), permissions);
        if (granted) {
            onAllPermissionsGranted(requestCode, Arrays.asList(permissions));
            return;
        }

        List<String> unGrantedPermissions = Utils.shouldShowRationale(this, permissions);
        if (unGrantedPermissions.size() > 0 && mRequest.getRationalCallback() != null) {
            mRequest.getRationalCallback().showRational(unGrantedPermissions, () -> requestPermissions(permissions, requestCode));
        } else {
            requestPermissions(permissions, requestCode);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode != mRequest.getRequestCode()) return;
        if (mRequest.getGrantedCallback() == null && mRequest.getDeniedCallback() == null) return;
        List<String> granted = new ArrayList<>();
        List<String> denied = new ArrayList<>();
        for (int i = 0; i < permissions.length; i++) {
            if (grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                granted.add(permissions[i]);
            } else {
                denied.add(permissions[i]);
            }
        }
        if (denied.size() > 0) {
            onAnyPermissionDenied(requestCode, denied);
        } else {
            onAllPermissionsGranted(requestCode, granted);
        }
    }

    private void onAllPermissionsGranted(int requestCode, List<String> permissions) {
        if (mRequest.getGrantedCallback() == null) return;
        mRequest.getGrantedCallback().onGranted(requestCode, permissions);
    }

    private void onAnyPermissionDenied(int requestCode, List<String> permissions) {
        if (mRequest.getDeniedCallback() == null) return;
        mRequest.getDeniedCallback().onDenied(requestCode, permissions, () ->
                getActivity().startActivityForResult(Utils.createSettingIntent(getActivity()), GO_SETTING_CODE));
    }

    private void setRequest(Request request) {
        this.mRequest = request;

    }
}

File: sloth/src/main/kotlin/com/xiansenliu/sloth/PuppetFragment.kt
package com.xiansenliu.sloth

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Fragment
import android.content.pm.PackageManager
import java.util.*

/**
 * Author       xinliu
 * Date         6/24/17
 * Time         3:08 PM
 * Used to request permissions and get the results
 */

class PuppetFragment : Fragment() {

    private lateinit var mRequest: Request

    @SuppressLint("NewApi")
    internal fun request(request: Request) {
        setRequest(request)
        val permissions = mRequest.permissions
        val requestCode = mRequest.requestCode

        val granted = arePermissionsGranted(activity, permissions)
        if (granted) {
            onAllPermissionsGranted(requestCode, permissions)
            return
        }

        val unGrantedPermissions = shouldShowRationale(this, permissions)
        if (unGrantedPermissions.isNotEmpty()) {
            mRequest.onRational.showRationale(unGrantedPermissions) { requestPermissions(permissions.toTypedArray(), requestCode) }
        } else {
            requestPermissions(permissions.toTypedArray(), requestCode)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode != mRequest.requestCode) return
        val granted = ArrayList<String>()
        val denied = ArrayList<String>()
        for (i in permissions.indices) {
            if (grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                granted.add(permissions[i])
            } else {
                denied.add(permissions[i])
            }
        }
        if (denied.size > 0) {
            onAnyPermissionDenied(requestCode, denied)
        } else {
            onAllPermissionsGranted(requestCode, granted)
        }
    }

    private fun onAllPermissionsGranted(requestCode: Int, permissions: List<String>) {
        mRequest.onGranted.onGranted(requestCode, permissions)
    }

    private fun onAnyPermissionDenied(requestCode: Int, permissions: List<String>) {
        mRequest.onDenied.onDenied(requestCode, permissions) { activity.startActivityForResult(createSettingIntent(activity), requestCode) }
    }

    private fun setRequest(request: Request) {
        this.mRequest = request
    }

    companion object {
        private val TAG = PuppetFragment::class.java.canonicalName
        fun create(act: Activity): PuppetFragment {
            var puppetFragment: PuppetFragment? = act.fragmentManager.findFragmentByTag(TAG) as PuppetFragment?
            if (puppetFragment == null) {
                puppetFragment = PuppetFragment()
                act.fragmentManager.beginTransaction()
                        .add(puppetFragment, TAG)
                        .commitAllowingStateLoss()
                act.fragmentManager.executePendingTransactions()
            }
            return puppetFragment
        }

    }
}
